### PR TITLE
Add EIP-5003: Insert Code into EOAs with `AUTHUSURP`

### PR DESCRIPTION
**Pull request by danfinlay**
Opening a draft for a proposed opcode that would allow publishing code at an EOA's address, converting it into a contract account, one way.

The specific proposed implementation is still under development, but this can serve as a starting point for discussion.

It is currently likely this will utilize the `AUTH` opcode as defined in 3074 (which will need to be split into an abstract-EIP), and then allow the authorized contract to perform a CREATE-like operation that target's the EOA's address for the new code publication.

### CODE REVIEW

**Comment by danfinlay** [comment]
I guess I need to wait to be assigned an EIP number before I can fix the bot's complaints.

File: EIPS/eip-draft_AUTH_USURP.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+---
+eip: <to be assigned>
```
Comment (lightclient) [inline EIPS/eip-draft_AUTH_USURP.md:2]:
```suggestion
eip: 5003
```

File: EIPS/eip-draft_AUTH_USURP.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+title: Replace Externally Owned Accounts with `AUTHUSURP`
+description: Allow migrating away from ECDSA by deploying code in place of an externally owned account.
+author: Dan Finlay (@danfinlay), Sam Wilson (@SamWilsn)
+discussions-to: <URL>
```
Comment (lightclient) [inline EIPS/eip-draft_AUTH_USURP.md:6]: Will need a `discussions-to` before merging.

File: EIPS/eip-draft_AUTH_USURP.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  - The initcode runs at the address `authorized`.
+  - If the initcode returns no bytes, its execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - After executing the initcode, but before the returned code is deployed, if the account's code is non-empty, the initcode's execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - The code is deployed into the account with the address `authorized`.
+
+## Rationale
+The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages.
+
+## Backwards Compatibility
+All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.
```
Comment (lightclient) [inline EIPS/eip-draft_AUTH_USURP.md:75]: Will need to fill out the relevant sections here and below. `Test Cases` and `Reference Implementation` are optional at this point so the blockers are `Rationale`, `Backwards Compatibility`, and `Security Considerations`.

**Review by lightclient — CHANGES_REQUESTED** [review]

**Review by Bezzat — APPROVED** [review]

File: EIPS/eip-draft_AUTH_USURP.md (hunk 2 above)
Comment (danfinlay) [reply EIPS/eip-draft_AUTH_USURP.md:6]:
```suggestion
discussions-to: [ethereum-magicians](https://ethereum-magicians.org/t/eip-5003-auth-usurp-publishing-code-at-an-eoa-address/8979)
```

File: EIPS/eip-draft_AUTH_USURP.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+---
+eip: 5003
+title: Replace Externally Owned Accounts with `AUTHUSURP`
+description: Allow migrating away from ECDSA by deploying code in place of an externally owned account.
+author: Dan Finlay (@danfinlay), Sam Wilson (@SamWilsn)
+discussions-to: [ethereum-magicians](https://ethereum-magicians.org/t/eip-5003-auth-usurp-publishing-code-at-an-eoa-address/8979)
```
Comment (lightclient) [inline EIPS/eip-draft_AUTH_USURP.md:6]:
```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-5003-auth-usurp-publishing-code-at-an-eoa-address/8979
```

**Review by lightclient — CHANGES_REQUESTED** [review]

File: EIPS/eip-draft_AUTH_USURP.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+## Backwards Compatibility
+All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.
+
+## Test Cases
+Test cases for an implementation are mandatory for EIPs that are affecting consensus changes.  If the test suite is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`.
+
+## Reference Implementation
+An optional section that contains a reference/example implementation that people can use to assist in understanding or implementing this specification.  If the implementation is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`.
+
```
Comment (MicahZoltu) [inline EIPS/eip-draft_AUTH_USURP.md:82]: These are optional sections, so it is best to leave them out until you have something for them.
```suggestion
```

File: EIPS/eip-draft_AUTH_USURP.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+## Security Considerations
+All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.
```
Comment (MicahZoltu) [inline EIPS/eip-draft_AUTH_USURP.md:84]:
```suggestion
TBD
```

File: EIPS/eip-draft_AUTH_USURP.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  - If `authorized` (as defined in EIP-3074) is unset, execution is invalid.
+  - If `authorized` points to an empty account, then `static_gas` remains 32,000. Otherwise, `static_gas` shall be 7,000.
+  - `AUTHUSURP` does not check the nonce of the `authorized` account.
+  - The initcode runs at the address `authorized`.
+  - If the initcode returns no bytes, its execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - After executing the initcode, but before the returned code is deployed, if the account's code is non-empty, the initcode's execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - The code is deployed into the account with the address `authorized`.
+
+## Rationale
+The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages.
```
Comment (MicahZoltu) [inline EIPS/eip-draft_AUTH_USURP.md:72]:
```suggestion
TBD
```

File: EIPS/eip-draft_AUTH_USURP.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+## Reference Implementation
+An optional section that contains a reference/example implementation that people can use to assist in understanding or implementing this specification.  If the implementation is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`.
+
+## Security Considerations
+All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.
+
+## Copyright
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```
Comment (MicahZoltu) [inline EIPS/eip-draft_AUTH_USURP.md:87]: This is a recent change to the template.
```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
While it would be best to fill in all sections before DRAFT, it isn't strictly required.  However, `TBD` stands out as more obviously incomplete than the placeholder text so that is preferred if you are going to choose to add those details later.

**Comment by danfinlay** [comment]
Changes accepted. Thanks, @MicahZoltu.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Filename needs to be changed to `EIPS/eip-5003.md` and the `discussions-to` link needs to be fixed (see comment from lightclient).

File: EIPS/eip-draft_AUTH_USURP.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+`AUTHUSURP` behaves identically to `CREATE` (`0xf0`), except as described below:
+
+  - If `authorized` (as defined in EIP-3074) is unset, execution is invalid.
+  - If `authorized` points to an empty account, then `static_gas` remains 32,000. Otherwise, `static_gas` shall be 7,000.
+  - `AUTHUSURP` does not check the nonce of the `authorized` account.
+  - The initcode runs at the address `authorized`.
+  - If the initcode returns no bytes, its execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - After executing the initcode, but before the returned code is deployed, if the account's code is non-empty, the initcode's execution frame must be reverted, and `AUTHUSURP` returns zero.
+  - The code is deployed into the account with the address `authorized`.
+
```
Comment (Pandapip1) [inline EIPS/eip-draft_AUTH_USURP.md:70]: Are transactions and signatures signed with the original private key still valid and able to be submitted to the blockchain, or should they revert?

**Review by Pandapip1 — APPROVED** [review]
I think this meets the basic requirements for a draft. Merging.

**Review by Pandapip1 — APPROVED** [review]

**Review by Mikh333 — APPROVED** [review]

**Pull request merged**